File: tree.java
public class tree {
    class Node{
        int data;
        Node right;
        Node left;

        Node(int data){
            this.data=data;
        }
    }

    public Node root=null;;

    public void buildTree(int data) {
        Node current=root;
        Node newNode=new Node(data);
        if(root==null){
root=newNode;
return;
        }
        while (true) {
            if (data>current.data) {
            if (current.right==null) {
                current.right=newNode;
                break;
            }
            else{
                current=current.right;
            }    
            }
            else{
                if (current.left==null) {
                    current.left=newNode;
                    break;
                }
                else{
                    current=current.left;
                }
            }
            
        }
        
    }

    public void remove(int data) {
        Node currentNode=root;
        Node parentNode=null;
        removeHelper(data,currentNode,parentNode);
    }

    public void removeHelper(int data,Node currentNode,Node parentNode) {
        while (currentNode!=null) {
            if (data>currentNode.data) {
                parentNode=currentNode;
                currentNode=currentNode.right;
            } else if (data<currentNode.data) {
                parentNode=currentNode;
                currentNode=currentNode.left;
            }
            else{
                if (currentNode.right!=null&&currentNode.left!=null) {
                    currentNode.data=getMinimum(currentNode.right);
                    removeHelper(currentNode.data, currentNode.right, currentNode);
                } 
                else {
                    if (parentNode==null) {
                        if (currentNode.left!=null) {
                            root=currentNode.left;
                        }
                        else{
                            root=currentNode.right;
                        }
                    }
                    else{
                        if (parentNode.left==null) {
                            if (currentNode.left==null) {
                                parentNode.right=currentNode.right;
                            }
                            else{
                                parentNode.right=currentNode.left;
                            }
                        }
                        else{
                            if (currentNode.left==null) {
                                System.out.println(parentNode.data);
                                parentNode.left=currentNode.right;
                            } else {
                                parentNode.left=currentNode.left;
                            }
                        }
                    }
                }

                return;
            }
        }
        
    }

    public int getMinimum(Node currentNode) {
        if (currentNode.left==null) {
            return currentNode.data;
        }
        else{
           return getMinimum(currentNode.left);
        }
    }

    public boolean contains(int data) {
        Node current=root;
        if (root==null) {
            return false;
        }
        while (current!=null) {
            if (data>current.data) {
                current=current.right;    
                }
                else if (data<current.data) {
                    current=current.left;
                } 
                else{
    return true;
                }
            }
        return false;
    }


    public void inOrderDisplay() {
        inOrderHelper(root);
    }

    public void postOrderDisplay() {
        postOrderHelper(root);
    }

    public void preOrderDisplay() {
        preOrderHelper(root);
    }


    public void inOrderHelper(Node node) {
        if (node!=null) {
            inOrderHelper(node.left);
            System.out.print(node.data+" ");
            inOrderHelper(node.right);
        }
    }

    public void preOrderHelper(Node node) {
        if (node!=null) {
            System.out.print(node.data+" ");
            preOrderHelper(node.left);
            preOrderHelper(node.right);
        }
        
    }
        public void postOrderHelper(Node node) {
            if (node!=null) {
                postOrderHelper(node.left);
                postOrderHelper(node.right);
                System.out.print(node.data+" ");
            }
            
        }
    public static void main(String[] args) {
        tree list=new tree();
        list.buildTree(10);
        list.buildTree(20);
        list.buildTree(8);
        list.buildTree(6);
        list.buildTree(7);
        list.buildTree(5);
        list.buildTree(9);
        list.buildTree(15);
        list.buildTree(25);
        list.buildTree(30);
        list.buildTree(35);
        list.buildTree(40);
        list.buildTree(45);
        list.buildTree(50);

        list.remove(25);
        System.out.println();
        list.inOrderDisplay();
        
    }
}

File: tree.js
class Node {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }
}

class Tree {
  constructor() {
    this.root = null;
  }

  buildTree(data) {
    const newNode = new Node(data);
    if (this.root === null) {
      this.root = newNode;
      return;
    }

    let current = this.root;
    while (true) {
      if (data > current.data) {
        if (current.right === null) {
          current.right = newNode;
          break;
        } else {
          current = current.right;
        }
      } else {
        if (current.left === null) {
          current.left = newNode;
          break;
        } else {
          current = current.left;
        }
      }
    }
  }

  remove(data) {
    this.root = this.removeHelper(data, this.root);
  }

  removeHelper(data, currentNode) {
    if (currentNode === null) {
      return null;
    }

    if (data > currentNode.data) {
      currentNode.right = this.removeHelper(data, currentNode.right);
    } else if (data < currentNode.data) {
      currentNode.left = this.removeHelper(data, currentNode.left);
    } else {
      if (currentNode.right !== null && currentNode.left !== null) {
        currentNode.data = this.getMinimum(currentNode.right);
        currentNode.right = this.removeHelper(currentNode.data, currentNode.right);
      } else {
        if (currentNode.left === null) {
          currentNode = currentNode.right;
        } else {
          currentNode = currentNode.left;
        }
      }
    }

    return currentNode;
  }

  getMinimum(currentNode) {
    if (currentNode.left === null) {
      return currentNode.data;
    } else {
      return this.getMinimum(currentNode.left);
    }
  }

  contains(data) {
    let current = this.root;
    while (current !== null) {
      if (data > current.data) {
        current = current.right;
      } else if (data < current.data) {
        current = current.left;
      } else {
        return true;
      }
    }
    return false;
  }

  inOrderDisplay() {
    this.inOrderHelper(this.root);
  }

  postOrderDisplay() {
    this.postOrderHelper(this.root);
  }

  preOrderDisplay() {
    this.preOrderHelper(this.root);
  }

  inOrderHelper(node) {
    if (node !== null) {
      this.inOrderHelper(node.left);
      console.log(node.data);
      this.inOrderHelper(node.right);
    }
  }

  preOrderHelper(node) {
    if (node !== null) {
      console.log(node.data);
      this.preOrderHelper(node.left);
      this.preOrderHelper(node.right);
    }
  }

  postOrderHelper(node) {
    if (node !== null) {
      this.postOrderHelper(node.left);
      this.postOrderHelper(node.right);
      console.log(node.data);
    }
  }
}

const tree = new Tree();
tree.buildTree(20);
tree.buildTree(5);
tree.buildTree(27);
tree.buildTree(3);
tree.buildTree(15);
tree.buildTree(25);
tree.buildTree(29);
tree.buildTree(7);
tree.buildTree(17);


tree.remove(5);
console.log("In-order Display:");
tree.postOrderDisplay();
